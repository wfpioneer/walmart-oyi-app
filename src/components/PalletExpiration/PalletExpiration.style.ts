import { StyleSheet } from 'react-native';
import COLOR from '../../themes/Color';

export default StyleSheet.create({
  headerText: {
    fontSize: 16,
    fontWeight: 'bold'
  },
  errorLabel: {
    color: COLOR.RED_500,
    textAlign: 'center'
  },
  effectiveDateHeaderItem: {
    fontSize: 15,
    color: COLOR.BLACK,
    textAlign: 'center'
  },
  effectiveDateContainer: {
    paddingHorizontal: 10,
    flexDirection: 'column',
    alignItems: 'center'
  },
  modifiedEffectiveDateContainer: {
    paddingHorizontal: 5,
    flexDirection: 'column',
    alignItems: 'center',
    backgroundColor: COLOR.YELLOW,
    borderWidth: 1
  }
});
