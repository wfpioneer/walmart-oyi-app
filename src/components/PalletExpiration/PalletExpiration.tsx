import React from 'react';
import { Platform, Text, TouchableOpacity, View } from 'react-native';
import DateTimePicker, { DateTimePickerEvent } from '@react-native-community/datetimepicker';
import { strings } from '../../locales';
import styles from './PalletExpiration.style';

export interface PalletExpirationProps {
  dateChanged: boolean;
  dateRemoved: boolean;
  expirationDate: string | undefined;
  newExpirationDate: string | undefined;
  minimumDate: Date | undefined;
  showPicker: boolean;
  setShowPicker: React.Dispatch<React.SetStateAction<boolean>>;
  onDateChange(event: DateTimePickerEvent, value: Date| undefined): void;
}

const getDateValue = (expirationDate: string | undefined, newExpirationDate: string | undefined): Date => {
  if (newExpirationDate) {
    return new Date(newExpirationDate)
  }

  if (expirationDate) {
    return new Date(expirationDate)
  }

  return new Date(Date.now());
};

export const PalletExpiration = (props: PalletExpirationProps): JSX.Element => {
  const {
    dateChanged,
    dateRemoved,
    expirationDate,
    newExpirationDate,
    minimumDate,
    onDateChange,
    showPicker,
    setShowPicker
  } = props;

  return (
    <View style={dateChanged ? styles.modifiedEffectiveDateContainer : styles.effectiveDateContainer}>
      <TouchableOpacity onPress={() => setShowPicker(true)} testID="openDate">
        <Text style={styles.headerText}>
          {strings('PALLET.EXPIRATION_DATE')}
        </Text>
        <Text style={(expirationDate || newExpirationDate || dateRemoved)
          ? styles.effectiveDateHeaderItem : styles.errorLabel}
        >
          {dateRemoved ? strings('GENERICS.REMOVED')
            : newExpirationDate || expirationDate || strings('GENERICS.REQUIRED')}
        </Text>
      </TouchableOpacity>
      {showPicker && (
        <DateTimePicker
          value={getDateValue(expirationDate, newExpirationDate)}
          mode="date"
          display={Platform.OS === 'ios' ? 'spinner' : 'default'}
          is24Hour={true}
          minimumDate={minimumDate}
          onChange={onDateChange}
          testID="datePicker"
        />
      )}
    </View>
  )
};

export default PalletExpiration;
