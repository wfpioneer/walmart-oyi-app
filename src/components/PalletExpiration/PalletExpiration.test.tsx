import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import PalletExpiration from './PalletExpiration';

describe('Tests rendering PalletExpiration', () => {
  let testExpirationDate: string | undefined = '05/31/2022';
  let testNewExpirationDate: string | undefined;
  const testSetShowPicker = jest.fn();
  const testDateChange = jest.fn();
  const testDate = new Date("10/17/1973");

  it('Renders the PalletExpiration with expiration date and no changes', () => {
    const { toJSON } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={false}
        setShowPicker={testSetShowPicker}
        dateRemoved={false}
        dateChanged={false}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the PalletExpiration with no expiration date and no changes', () => {
    testExpirationDate = undefined;

    const { toJSON } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={false}
        setShowPicker={testSetShowPicker}
        dateRemoved={false}
        dateChanged={false}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the PalletExpiration with newExpiration date', () => {
    testNewExpirationDate = '05/31/2022';

    const { toJSON } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={false}
        setShowPicker={testSetShowPicker}
        dateRemoved={false}
        dateChanged={true}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the PalletExpiration with date removed', () => {
    testNewExpirationDate = undefined;

    const { toJSON } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={false}
        setShowPicker={testSetShowPicker}
        dateRemoved={true}
        dateChanged={true}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the PalletExpiration showing picker', () => {
    testExpirationDate = '5/30/2002';
    testNewExpirationDate = undefined;

    const { toJSON } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={true}
        setShowPicker={testSetShowPicker}
        dateRemoved={true}
        dateChanged={true}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('exercise onClick and onDateChange', () => {
    testExpirationDate = '5/30/2002';
    testNewExpirationDate = undefined;

    const { toJSON, getByTestId } = render(
      <PalletExpiration
        expirationDate={testExpirationDate}
        newExpirationDate={testNewExpirationDate}
        showPicker={true}
        setShowPicker={testSetShowPicker}
        dateRemoved={true}
        dateChanged={true}
        onDateChange={testDateChange}
        minimumDate={testDate}
      />
    );
    const testClick = getByTestId('openDate');
    const testChange = getByTestId('datePicker');
    fireEvent.press(testClick);
    expect(testSetShowPicker).toBeCalledTimes(1);
    fireEvent(testChange, 'onChange', { nativeEvent: { timestamp: '01/01/1976' } });
    expect(testDateChange).toBeCalledTimes(1);
  });
});
