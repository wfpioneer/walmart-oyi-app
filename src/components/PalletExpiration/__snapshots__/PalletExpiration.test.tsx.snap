// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests rendering PalletExpiration Renders the PalletExpiration showing picker 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFE0",
      "borderWidth": 1,
      "flexDirection": "column",
      "paddingHorizontal": 5,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="openDate"
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      [missing "en.PALLET.EXPIRATION_DATE" translation]
    </Text>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      [missing "en.GENERICS.REMOVED" translation]
    </Text>
  </View>
  <RNDateTimePicker
    date={1022734800000}
    displayIOS="spinner"
    enabled={true}
    minimumDate={119682000000}
    mode="date"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 216,
      }
    }
    testID="datePicker"
  />
</View>
`;

exports[`Tests rendering PalletExpiration Renders the PalletExpiration with date removed 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFE0",
      "borderWidth": 1,
      "flexDirection": "column",
      "paddingHorizontal": 5,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="openDate"
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      [missing "en.PALLET.EXPIRATION_DATE" translation]
    </Text>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      [missing "en.GENERICS.REMOVED" translation]
    </Text>
  </View>
</View>
`;

exports[`Tests rendering PalletExpiration Renders the PalletExpiration with expiration date and no changes 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "paddingHorizontal": 10,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="openDate"
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      [missing "en.PALLET.EXPIRATION_DATE" translation]
    </Text>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      05/31/2022
    </Text>
  </View>
</View>
`;

exports[`Tests rendering PalletExpiration Renders the PalletExpiration with newExpiration date 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFE0",
      "borderWidth": 1,
      "flexDirection": "column",
      "paddingHorizontal": 5,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="openDate"
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      [missing "en.PALLET.EXPIRATION_DATE" translation]
    </Text>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      05/31/2022
    </Text>
  </View>
</View>
`;

exports[`Tests rendering PalletExpiration Renders the PalletExpiration with no expiration date and no changes 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "paddingHorizontal": 10,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="openDate"
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      [missing "en.PALLET.EXPIRATION_DATE" translation]
    </Text>
    <Text
      style={
        Object {
          "color": "#f44336",
          "textAlign": "center",
        }
      }
    >
      [missing "en.GENERICS.REQUIRED" translation]
    </Text>
  </View>
</View>
`;
